### ISSUE TITLE
Branching yaml install files

### ISSUE THREAD

**Issue by jqmichael** [kubernetes-sigs/aws-efs-csi-driver#207]
**Is your feature request related to a problem?/Why is this needed**

/feature

**Describe the solution you'd like in detail**

Today, the `node.yaml` file under `deploy` directory is shared between `dev` and `stable` overlays. Any updates to `node.yaml` will be reflected in both overlays. Users are instructed to use the `stable` version of the driver with the `kubectl apply -k "github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=master" ` command. Therefore, any changes that are only meant for the `dev` overlay are getting picked up by `stable` overlay and caused customer impact like https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/191. Kustomize, though powerful in many aspects, isn't designed to version yaml files. 

Proposal:
Reconstruct the yaml install directory to the following structure:

```
deploy
         k8s
             dev
             v0.3.0
             v0.4.0
```

That way, once a new release is cut, users can `kubectl apply` to the folder for the specific version. And all the files under each release folder won't be updated unless there's a bug fix that needs to be backfilled. 

The existing `deploy/kubernetes` will get deprecated and remain as is for backwards compatibility. New changes to the yaml files will only be made to the new folder.  

Other changes:
README to instruct users to switch to the new folder and deprecate the old one.
Change test workflow to pick up latest yaml file from new folder. 

**Describe alternatives you've considered**

We could rely on git tags as an alternative for versioning. But tags are immutable, so any bug fixes can only be made to new tags, which isn't flexible. 

**Additional context**

**Comment by jqmichael** [comment]
@2uasimojo @leakingtapan @ossareh

**Comment by 2uasimojo** [comment]
Love this @jqmichael. It ought to help significantly with tagged releases.

There's still a hole, which is that any untagged overlay (call it `dev`, `stable`, `latest`, whatever) is still a moving target. For example, if you set up your driver pod using the `dev` overlay from the master branch which refers to the `latest` image tag, and then the `latest` image changes, if you bounce your DaemonSet, you'll end up with the new image but the old YAML, as described in #192.

The way around this is to make sure you're always locking into a specific commit level for *both* the YAML and the image. Unfortunately it would be tough to do this in an in-repo overlay, because the overlay would have to refer to the image tag corresponding to the commit in which the overlay is tracked!

What we could do is provide a `make` target that generates/edits the overlay for you. For tagged releases the process stays the same; for untagged levels you would need to have a clone of the repo (or edit by hand).

Am I making sense here?

**Comment by ossareh** [comment]
@2uasimojo yah, that makes sense. It certainly [helps me understand why I saw my docker sha change unexpectedly](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/195#issuecomment-650280097) two weeks ago; i.e. the DaemonSet changing triggered fetching the incompatible latest sha. It also seems inevitable as part of using mutable docker tags. 

To break the problem into it's constituent parts:

 1. infrequent releases encourage the use of mutable tags like `latest`
 2. mutable tags expose tight coupling between source and config

RE: (1) #202 goes a long way toward improving that situation. I can see how @jqmichael solution builds on that and further improves the situation.

RE: (2) perhaps it's just a case of being more explicit in the [development documentation](https://github.com/kubernetes-sigs/aws-efs-csi-driver#development)? If you're using a mutable tag like `dev` or `latest` you should be made aware of this coupling and given instructions on how to mitigate it (see: https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/192#issuecomment-652855976). Any tooling that makes this easier is great, including running a script to generate inputs to the `kubectl apply -k` phase.

**Comment by leakingtapan** [comment]
IMHO, we don't have to explicitly list out all the versions under `deploy` directory just to version the manifest (I actually used this pattern during initial development of ebs csi driver). We could just rely on git tagging to version the manifest (as this is what is means). Since by the time a release happens, that's the file we want. Something like below is already working:

```
# deploy v0.3.0 driver do:
kubectl apply -k "github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=v0.3.0"

# deploy v0.4.0 driver do:
kubectl apply -k "github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=v0.4.0"

# deploy master driver do:
kubectl apply -k "github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=master"
```

So that we just need to fix documentation to say:
* master branch is always unstable, good to use for dev environment, never use for production.
* use the version tag for kubestomize

**Comment by jqmichael** [comment]
> So that we just need to fix documentation to say:
master branch is always unstable, good to use for dev environment, never use for production.
use the version tag for kubestomize

I think we should do better than purely relying on documentation. 

There're a couple of reasons I think branching the yaml file in maser branch is preferable. 

1) Backwards compatible for existing users who rely on 
`kubectl apply -k "github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=master"` (that has been the case for a long time). 
No matter what the latest doc says, most existing users will keep using the previously provided command to install the driver from master branch. I think this is the biggest motivation for deprecating this pattern going forward.

2) Avoid coupling bug fixes in the yaml file with immutable tags and releases. 
If there's a bug introduced in the yaml file when cutting a release, bug fixes won't be blocked by project admins cutting a new release. The speed of getting into market would be significantly faster. 

Thoughts?

@2uasimojo @leakingtapan @ossareh

**Comment by leakingtapan** [comment]
> Backwards compatible for existing users who rely on
kubectl apply -k "github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=master" (that has been the case for a long time).
No matter what the latest doc says, many existing users will keep using the previously provided command to install the driver from master branch (it might even be integrated in their CI/CD system). I think this is the biggest motivation for deprecating this pattern going forward.

Actually only v0.3 is released with kustomize. So I don’t think it’s out for too long. I also don’t see the compatibility concern, since it’s just like docket image latest tag, it exists but is discouraged for prod usage.

> Avoid coupling bug fixes in the yaml file with immutable tags and releases.
If there's a bug introduced in the yaml file when cutting a release, bug fixes won't be blocked by project admins cutting a new release. The speed of getting into market would be significantly faster.

Well. This also means any accidental changes that got merged could also be picked up before any proper testing or release is done for the branched yaml. And that is exactly what a release process aims to solve. How do you make sure every change to the branched yaml is tested with its corresponding k8s version(s)? I There is no such assurance once yaml is branched, but the release branch with tag does.

That being said I would rather invest in testing and release automation instead of branching yaml (which indirectly discourages releases, since if all the problems could be solved without release, why bother doing release). If we do it right, we could even auto cut the release when x conditions are met.

Last but not least, I think @wongma7 already has release permission. If you care about certain fixes, you should be able to do it already. And it’s just several hours’ work, as some work is already automated. I would rather spending time on the release than debating workarounds like this one.

**Comment by jqmichael** [comment]
> I also don’t see the compatibility concern, since it’s just like docker image latest tag, it exists but is discouraged for prod usage

Are we willing to break users who follows 
```
kubectl apply -k "github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=master
```
as suggested by the README doc as well as EKS doc going forward? 
https://docs.aws.amazon.com/eks/latest/userguide/efs-csi.html 

And for uses who were impacted by bugs like https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/191, what's our suggestion to them? Don't follow the old doc, start downloading from a tagged branch since the old doc was incorrect? This is the backwards incompatible change I was referring to.

**Comment by wongma7** [comment]
First of all: we must have git branching+tagging. Git branching is necessary for maintaining stable release branch(es) alongside an unstable master branch. Without git branching we cannot do bugfix releases. Tagging is how you notify users there is a new release. We have no other way to tell them. So I agree with @leakingtapan: with git branches+tags, we can and should change the instructions to point to `release-X.Y` git branches and/or `vX.Y.Z` git tags:
`kubectl apply -k "...?ref=release-0.3` to get the latest v0.3.Z release.
`kubectl apply -k "...?ref=v0.3.0` to get the specific v0.3.0 release.

Now, what to do with the stable/dev overlays given that we ran into https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/191?

Remember the original goal of the overlays is to have stable point to the newest release image+newest release's YAMLs and dev to master image+master YAMLs.

Option 1 is to remove+deprecate them and replace them with something else (@jqmichael). When we are about to make a release, we need to make a new directory and copy the current dev YAMLs into it.

Option 2 is to slightly change the meaning of stable in master branch to point to `latest` but maintain the meaning of stable in release branches (@leakingtapan). When we are about to make a release, we don't need to do anything, the stable overlay in master branch becomes the stable overlay in release-0.x branch.

Option 3 is to keep them as is. When we are about to make a release, we need to take the patches in the dev overlay and apply them to the base YAMLs so that the stable overlay gets "patched" too! <- MY PREFERRED SOLUTION. We do not have to do anything today. But the next time someone makes a PR intending to change the YAMLs they must add patches to dev instead of editing the base YAMLs directly.

https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/191 does not prescribe that we remove/deprecate/change the stable+dev kustomize overlays. The lesson to be learned there is: having a patch that patches *only* the image field can be insufficient. Meaning you may need additional patches to differentiate dev from stable. So it's 100% achievable to have stable+dev kustomize overlays and maintain the stability of the stable overlay in master branch. For https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/191 in particular, I think what we could have done is added a patch to overlays/dev to add the new hostPath volume. And once we are ready to release 0.4.0, take the patch in overlays/dev and apply it to base/node.yaml. Any reason this wouldn't work?

> I would rather spending time on the release than debating workarounds like this one.

Agree. That is why I propose option 3, do nothing...

**Comment by 2uasimojo** [comment]
Option 3 makes sense to me, as long as we beef up the docs. Suggest:
- Update docs (both [here](https://github.com/kubernetes-sigs/aws-efs-csi-driver/blob/778131e6fdfce466bbada3fb08b1e5bdd50c072b/docs/README.md#installation) and [eks](https://docs.aws.amazon.com/eks/latest/userguide/efs-csi.html) to describe how to apply a specific release (`kubectl apply -k ...?ref=$tag`)
- Drop a comment in the base YAMLs warning not to edit except when cutting a release
- Drop a comment in the overlays reminding to fold them into the base when cutting a release
- Assuming there is documentation somewhere for how to cut a release, make sure it includes :arrow_up: (whereupon :arrow_up:  could mainly refer to said documentation)
- Add some words in the [Development section](https://github.com/kubernetes-sigs/aws-efs-csi-driver/blob/778131e6fdfce466bbada3fb08b1e5bdd50c072b/docs/README.md#development) of the docs describing how to use commit hashes if you need to lock into a specific dev level.

**Comment by jqmichael** [comment]
> Update docs (both here and eks to describe how to apply a specific release (kubectl apply -k ...?ref=$tag)

Agreed we should update our doc no matter what.

Personally, I think option3 significantly adds complexity into future updates to the yaml file. Imagine if we need to consistently add, delete and update the yaml structure, and purely relying on manual json patch could leads to a series of nightmare and bugs. 

Any reason maintaining two copies of yaml files for dev and stable version is unacceptable? It doesn't sound like the end of the world if we duplicate two yaml files. 

@wongma7 @2uasimojo

**Comment by 2uasimojo** [comment]
@jqmichael I don't have a strong opinion on that front. Option 3 is acceptable to me, but so are others, including:

**Option 4:** Get rid of the overlays entirely.
- Always use `?ref=$tag_or_commit` to apply what you want.
- You're always guaranteed frozen if you use a `$commit`, but
   a) that's unsupported dev-only unstable yatta yatta; and
   b) you are responsible for overriding the image reference with the same hash (manually or via some script as [suggested](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/207#issuecomment-653005234)).
- Tags represent branches in which the (one and only) node.yaml refers to the corresponding identical image tag. The process of branching a full release entails:
  - editing the image tag in node.yaml to correspond to the branch name;
  - tagging the corresponding image with the same;
  - updating the `stable` branch to point to the new tagged branch.
- Some tags (e.g. `v$M.$m.$p`) are frozen forever. These branches should be closed and new commits forbidden. If bugs need to be fixed or whatever, they need to go into a dev sub-branch (e.g. `v$M.$m-dev`) which would eventually be released as `v$M.$m.${p+1}` or similar.
- Other tags are moving targets of varying levels of volatility, which should be clearly documented. Committing to these branches entails nothing special. Examples:
  - `$M.$m-dev` as noted above represents in-flight development on top of the latest `v$M.$m.$p` and therefore changes e.g. any time a fix is backported. `$M.$m-dev` happens to be identical to the latest `v$M.$m.$p` immediately after that release.
  - `stable` is always in sync with the latest `v$M.$m.$p`, so it moves any time a new full release happens.
  - `latest` points to master. (Though IMO we should get rid of `latest` as an image tag and use `master` instead so we can keep the paradigm of image tag corresponding to git branch.)

**Comment by jqmichael** [comment]
Thanks @2uasimojo. That's very good call out. 

Looks like there are enough push back to completely restructure the existing overlay hierarchy.  

What if we meet in the middle ground by going with a slightly modified version of option 3? To be specific, reuse the existing install mechanism, but move the `node.yaml` file to each overlay folder to avoid the manual json patch headache? 

Something like 

```
base
  csidrvier.yaml
  kustomization.yaml
overlays
  dev
    node.yaml
    kustomization.yaml     
  stable
    node.yaml
    kustomization.yaml
```

The stable folder will only be updated in the release process as mentioned by @2uasimojo at https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/207#issuecomment-655056328

Thoughts?

@ossareh @leakingtapan @2uasimojo @wongma7

**Comment by 2uasimojo** [comment]
What will the image tag be set to in those overlays? For `dev/node.yaml` presumably `latest` (or `master` if you take my suggestion above). But for `stable/node.yaml`? It would feel pretty ooky to me to have `master`'s `stable/node.yaml` hardcoded to point to a `v$M.$m.$p` image tag.

And no matter which way you slice it, `stable` on `master` is still a moving target, and subject to the [sync problem](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/207#issuecomment-653005234), and therefore not very... stable.

So IMO nothing is actually gained by having overlays at all *unless* they're specifically named after (and frozen at) tags (your original suggestion @jqmichael). But if I may paraphrase/interpret @leakingtapan [above](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/207#issuecomment-653388592), it doesn't make sense to have hardcoded paths in the repository that correspond to tags when you can just... use the tags. That's what they're for.

To summarize:
There is no change to yaml or documentation that's going to make `master`'s `stable` actually *be* stable. So anyone [relying on the current docs](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/207#issuecomment-654554555) is already broken, but [stealthily](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/207#issuecomment-653005234). So it's actually _better_ to break them hard and force them to do something that works as expected. So I'm in favor of getting rid of the `stable` overlay for sure; and the `dev` overlay stops making sense in a release branch, so we should get rid of that too.

**Comment by jqmichael** [comment]
Unfortunately, updating doc to point to 

`github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=0.3.0` won't work because the images didn't get updated when cutting the 0.3.0 release. 
https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/172

We have to live with it until the next release before updating our doc. :( 

@leakingtapan 
What's your thoughts moving to a model outlined at https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/207#issuecomment-655120402 by @2uasimojo ? Essentially, merge dev and stable overlay and update doc to ask user to apply a tagged kustomized directory going forward?

**Comment by leakingtapan** [comment]
> What's your thoughts moving to a model outlined at #207 (comment) by @2uasimojo in the long term? Essentially, merge dev and stable overlay and update doc to ask user to apply a tagged kustomized directory?

@2uasimojo's suggestion sounds good to me. And I think breaking end user's current usage of `ref=master` kustomize apply and switch to use `ref={tag}` is the right thing to do.

In long term, I would consider this problem along with the sync problem between helm chart. Since once we get to a state where kustomize files are properly versioned and layered, we'll come into how to sync it with the helm chart. And these two could get out of sync fairly easy. So instead, we could focus on auto-generating the kustomize files from helm chart, and kustomize files are managed by the generation tool.
